File: fluxc/src/main/java/org/wordpress/android/fluxc/persistence/PluginSqlUtils.java
package org.wordpress.android.fluxc.persistence;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;

import com.wellsql.generated.PluginDirectoryModelTable;
import com.wellsql.generated.SitePluginModelTable;
import com.wellsql.generated.WPOrgPluginModelTable;
import com.yarolegovich.wellsql.WellSql;

import org.wordpress.android.fluxc.model.SiteModel;
import org.wordpress.android.fluxc.model.plugin.PluginDirectoryModel;
import org.wordpress.android.fluxc.model.plugin.PluginDirectoryType;
import org.wordpress.android.fluxc.model.plugin.SitePluginModel;
import org.wordpress.android.fluxc.model.plugin.WPOrgPluginModel;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;

import static com.yarolegovich.wellsql.SelectQuery.ORDER_ASCENDING;

public class PluginSqlUtils {
    public static @NonNull List<SitePluginModel> getSitePlugins(@NonNull SiteModel site) {
        return WellSql.select(SitePluginModel.class)
                .where()
                .equals(SitePluginModelTable.LOCAL_SITE_ID, site.getId())
                .endWhere()
                .orderBy(SitePluginModelTable.DISPLAY_NAME, ORDER_ASCENDING)
                .getAsModel();
    }

    public static void insertOrReplaceSitePlugins(@NonNull SiteModel site, @NonNull List<SitePluginModel> plugins) {
        // Remove previous plugins for this site
        removeSitePlugins(site);
        // Insert new plugins for this site
        for (SitePluginModel sitePluginModel : plugins) {
            sitePluginModel.setLocalSiteId(site.getId());
        }
        WellSql.insert(plugins).asSingleTransaction(true).execute();
    }

    private static void removeSitePlugins(@NonNull SiteModel site) {
        WellSql.delete(SitePluginModel.class)
                .where()
                .equals(SitePluginModelTable.LOCAL_SITE_ID, site.getId())
                .endWhere().execute();
    }

    public static int insertOrUpdateSitePlugin(SiteModel site, SitePluginModel plugin) {
        if (plugin == null) {
            return 0;
        }

        SitePluginModel oldPlugin = getSitePluginBySlug(site, plugin.getSlug());
        plugin.setLocalSiteId(site.getId()); // Make sure the site id is set (if the plugin is retrieved from network)
        if (oldPlugin == null) {
            WellSql.insert(plugin).execute();
            return 1;
        } else {
            int oldId = oldPlugin.getId();
            return WellSql.update(SitePluginModel.class).whereId(oldId)
                    .put(plugin, new UpdateAllExceptId<>(SitePluginModel.class)).execute();
        }
    }

    public static int deleteSitePlugins(@NonNull SiteModel site) {
        return WellSql.delete(SitePluginModel.class)
                .where()
                .equals(SitePluginModelTable.LOCAL_SITE_ID, site.getId())
                .endWhere().execute();
    }

    public static int deleteSitePlugin(SiteModel site, String slug) {
        if (TextUtils.isEmpty(slug)) {
            return 0;
        }
        // The local id of the plugin might not be set if it's coming from a network request,
        // using site id and slug is a safer approach here
        return WellSql.delete(SitePluginModel.class)
                .where()
                .equals(SitePluginModelTable.SLUG, slug)
                .equals(SitePluginModelTable.LOCAL_SITE_ID, site.getId())
                .endWhere().execute();
    }

    public static SitePluginModel getSitePluginBySlug(@NonNull SiteModel site, String slug) {
        List<SitePluginModel> result = WellSql.select(SitePluginModel.class)
                .where().equals(SitePluginModelTable.SLUG, slug)
                .equals(SitePluginModelTable.LOCAL_SITE_ID, site.getId())
                .endWhere().getAsModel();
        return result.isEmpty() ? null : result.get(0);
    }

    public static @Nullable WPOrgPluginModel getWPOrgPluginBySlug(String slug) {
        List<WPOrgPluginModel> result = WellSql.select(WPOrgPluginModel.class)
                .where().equals(WPOrgPluginModelTable.SLUG, slug)
                .endWhere().getAsModel();
        return result.isEmpty() ? null : result.get(0);
    }

    public static @NonNull List<WPOrgPluginModel> getWPOrgPluginsForDirectory(PluginDirectoryType directoryType) {
        List<PluginDirectoryModel> directoryModels = getPluginDirectoriesForType(directoryType);
        if (directoryModels.size() == 0) {
            // No directories found, return an empty list
            return new ArrayList<>();
        }
        List<String> slugList = new ArrayList<>(directoryModels.size());
        final HashMap<String, Integer> orderMap = new HashMap<>();
        for (int i = 0; i < directoryModels.size(); i++) {
            String slug = directoryModels.get(i).getSlug();
            slugList.add(slug);
            orderMap.put(slug, i);
        }
        List<WPOrgPluginModel> wpOrgPlugins = WellSql.select(WPOrgPluginModel.class)
                .where().isIn(WPOrgPluginModelTable.SLUG, slugList)
                .endWhere().getAsModel();
        // We need to manually order the list according to the directory models since SQLite will return mixed results
        Collections.sort(wpOrgPlugins, new Comparator<WPOrgPluginModel>() {
            @Override
            public int compare(WPOrgPluginModel plugin1, WPOrgPluginModel plugin2) {
                return orderMap.get(plugin1.getSlug()).compareTo(orderMap.get(plugin2.getSlug()));
            }
        });
        return wpOrgPlugins;
    }

    public static int insertOrUpdateWPOrgPlugin(WPOrgPluginModel wpOrgPluginModel) {
        if (wpOrgPluginModel == null) {
            return 0;
        }

        // Slug is the primary key in remote, so we should use that to identify WPOrgPluginModels
        WPOrgPluginModel oldPlugin = getWPOrgPluginBySlug(wpOrgPluginModel.getSlug());

        if (oldPlugin == null) {
            WellSql.insert(wpOrgPluginModel).execute();
            return 1;
        } else {
            int oldId = oldPlugin.getId();
            return WellSql.update(WPOrgPluginModel.class).whereId(oldId)
                    .put(wpOrgPluginModel, new UpdateAllExceptId<>(WPOrgPluginModel.class)).execute();
        }
    }

    public static int insertOrUpdateWPOrgPluginList(List<WPOrgPluginModel> wpOrgPluginModels) {
        if (wpOrgPluginModels == null) {
            return 0;
        }

        int result = 0;
        for (WPOrgPluginModel pluginModel : wpOrgPluginModels) {
            result += insertOrUpdateWPOrgPlugin(pluginModel);
        }
        return result;
    }

    // Plugin Directory

    public static void deletePluginDirectoryForType(PluginDirectoryType directoryType) {
        WellSql.delete(PluginDirectoryModel.class)
                .where()
                .equals(PluginDirectoryModelTable.DIRECTORY_TYPE, directoryType.toString())
                .endWhere().execute();
    }

    public static void insertPluginDirectoryList(@Nullable List<PluginDirectoryModel> pluginDirectories) {
        if (pluginDirectories == null) {
            return;
        }

        WellSql.insert(pluginDirectories).asSingleTransaction(true).execute();
    }

    public static int getLastRequestedPageForDirectoryType(PluginDirectoryType directoryType) {
        List<PluginDirectoryModel> list = getPluginDirectoriesForType(directoryType);
        int page = 0;
        for (PluginDirectoryModel pluginDirectoryModel : list) {
            page = Math.max(pluginDirectoryModel.getPage(), page);
        }
        return page;
    }

    private static @NonNull List<PluginDirectoryModel> getPluginDirectoriesForType(PluginDirectoryType directoryType) {
        return WellSql.select(PluginDirectoryModel.class)
                .where()
                .equals(PluginDirectoryModelTable.DIRECTORY_TYPE, directoryType)
                .endWhere()
                .getAsModel();
    }
}

File: fluxc/src/main/java/org/wordpress/android/fluxc/persistence/PluginSqlUtils.kt
package org.wordpress.android.fluxc.persistence

import android.text.TextUtils
import com.wellsql.generated.PluginDirectoryModelTable
import com.wellsql.generated.SitePluginModelTable
import com.wellsql.generated.WPOrgPluginModelTable
import com.yarolegovich.wellsql.SelectQuery
import com.yarolegovich.wellsql.WellSql
import org.wordpress.android.fluxc.model.SiteModel
import org.wordpress.android.fluxc.model.plugin.PluginDirectoryModel
import org.wordpress.android.fluxc.model.plugin.PluginDirectoryType
import org.wordpress.android.fluxc.model.plugin.SitePluginModel
import org.wordpress.android.fluxc.model.plugin.WPOrgPluginModel
import kotlin.math.max

object PluginSqlUtils {
    @JvmStatic
    fun getSitePlugins(site: SiteModel): List<SitePluginModel> =
        WellSql.select(SitePluginModel::class.java)
            .where()
            .equals(SitePluginModelTable.LOCAL_SITE_ID, site.id)
            .endWhere()
            .orderBy(SitePluginModelTable.DISPLAY_NAME, SelectQuery.ORDER_ASCENDING)
            .asModel

    @JvmStatic
    fun insertOrReplaceSitePlugins(site: SiteModel, plugins: List<SitePluginModel>) {
        // Remove previous plugins for this site
        removeSitePlugins(site)
        // Insert new plugins for this site
        for (sitePluginModel in plugins) {
            sitePluginModel.localSiteId = site.id
        }
        WellSql.insert(plugins).asSingleTransaction(true).execute()
    }

    private fun removeSitePlugins(site: SiteModel) =
        WellSql.delete(SitePluginModel::class.java)
            .where()
            .equals(SitePluginModelTable.LOCAL_SITE_ID, site.id)
            .endWhere().execute()

    @JvmStatic
    fun insertOrUpdateSitePlugin(site: SiteModel, plugin: SitePluginModel?): Int {
        if (plugin == null) {
            return 0
        }
        val oldPlugin = getSitePluginBySlug(site, plugin.slug)
        // Make sure the site id is set (if the plugin is retrieved from network)
        plugin.localSiteId = site.id
        return if (oldPlugin == null) {
            WellSql.insert(plugin).execute()
            1
        } else {
            val oldId = oldPlugin.id
            WellSql.update(SitePluginModel::class.java)
                .whereId(oldId)
                .put(plugin, UpdateAllExceptId(SitePluginModel::class.java))
                .execute()
        }
    }

    @JvmStatic
    fun deleteSitePlugins(site: SiteModel): Int =
        WellSql.delete(SitePluginModel::class.java)
            .where()
            .equals(SitePluginModelTable.LOCAL_SITE_ID, site.id)
            .endWhere().execute()

    @JvmStatic
    fun deleteSitePlugin(site: SiteModel, slug: String?): Int {
        return if (TextUtils.isEmpty(slug)) {
            0
        } else {
            // The local id of the plugin might not be set if it's coming from a network request,
            // using site id and slug is a safer approach here
            WellSql.delete(SitePluginModel::class.java)
                .where()
                .equals(SitePluginModelTable.SLUG, slug)
                .equals(SitePluginModelTable.LOCAL_SITE_ID, site.id)
                .endWhere().execute()
        }
    }

    @JvmStatic
    fun getSitePluginBySlug(site: SiteModel, slug: String?): SitePluginModel? {
        val result = WellSql.select(SitePluginModel::class.java).where()
            .equals(SitePluginModelTable.SLUG, slug)
            .equals(SitePluginModelTable.LOCAL_SITE_ID, site.id)
            .endWhere()
            .asModel
        return if (result.isEmpty()) null else result.first()
    }

    fun getSitePluginByName(site: SiteModel, pluginName: String?): SitePluginModel? {
        val result = WellSql.select(SitePluginModel::class.java)
            .where()
            .equals(SitePluginModelTable.NAME, pluginName)
            .equals(SitePluginModelTable.LOCAL_SITE_ID, site.id)
            .endWhere()
            .asModel
        return if (result.isEmpty()) null else result.first()
    }

    fun getSitePluginByNames(site: SiteModel, pluginNames: List<String?>?): List<SitePluginModel> =
        WellSql.select(SitePluginModel::class.java)
            .where()
            .isIn(SitePluginModelTable.NAME, pluginNames)
            .equals(SitePluginModelTable.LOCAL_SITE_ID, site.id)
            .endWhere()
            .asModel

    @JvmStatic
    fun getWPOrgPluginBySlug(slug: String?): WPOrgPluginModel? {
        val result = WellSql.select(WPOrgPluginModel::class.java)
            .where()
            .equals(WPOrgPluginModelTable.SLUG, slug)
            .endWhere()
            .asModel
        return if (result.isEmpty()) null else result.first()
    }

    @JvmStatic
    fun getWPOrgPluginsForDirectory(directoryType: PluginDirectoryType?): List<WPOrgPluginModel> {
        val directoryModels = getPluginDirectoriesForType(directoryType)
        if (directoryModels.isEmpty()) {
            // No directories found, return an empty list
            return ArrayList()
        }
        val slugList = ArrayList<String?>(directoryModels.size)
        val orderMap = HashMap<String, Int>()
        directoryModels.forEachIndexed { index, pluginDirectoryModel ->
            val slug = pluginDirectoryModel.slug
            slugList.add(slug)
            orderMap[slug] = index
        }

        val batches = slugList.chunked(WellSqlConfig.SQLITE_MAX_VARIABLE_NUMBER)
        val wpOrgPlugins = mutableListOf<WPOrgPluginModel>()
        batches.forEach {
            val batchQueryResult = WellSql.select(WPOrgPluginModel::class.java)
                .where()
                .isIn(WPOrgPluginModelTable.SLUG, it)
                .endWhere()
                .asModel
            wpOrgPlugins.addAll(batchQueryResult)
        }
        // We need to manually order the list according to the directory models since SQLite will
        // return mixed results
        wpOrgPlugins.sortWith { plugin1, plugin2 ->
            val order1 = orderMap[plugin1.slug] ?: 0
            val order2 = orderMap[plugin2.slug] ?: 0
            order1.compareTo(order2)
        }
        return wpOrgPlugins
    }

    @JvmStatic
    fun insertOrUpdateWPOrgPlugin(wpOrgPluginModel: WPOrgPluginModel?): Int {
        if (wpOrgPluginModel == null) {
            return 0
        }

        // Slug is the primary key in remote, so we should use that to identify WPOrgPluginModels
        val oldPlugin = getWPOrgPluginBySlug(wpOrgPluginModel.slug)

        return if (oldPlugin == null) {
            WellSql.insert(wpOrgPluginModel).execute()
            1
        } else {
            val oldId = oldPlugin.id
            WellSql.update(WPOrgPluginModel::class.java)
                .whereId(oldId)
                .put(wpOrgPluginModel, UpdateAllExceptId(WPOrgPluginModel::class.java))
                .execute()
        }
    }

    @JvmStatic
    fun insertOrUpdateWPOrgPluginList(wpOrgPluginModels: List<WPOrgPluginModel?>?): Int {
        if (wpOrgPluginModels == null) {
            return 0
        }

        var result = 0
        wpOrgPluginModels.forEach { result += insertOrUpdateWPOrgPlugin(it) }
        return result
    }

    // Plugin Directory

    @JvmStatic
    fun deletePluginDirectoryForType(directoryType: PluginDirectoryType) =
        WellSql.delete(PluginDirectoryModel::class.java)
            .where()
            .equals(PluginDirectoryModelTable.DIRECTORY_TYPE, directoryType.toString())
            .endWhere()
            .execute()

    @JvmStatic
    fun insertPluginDirectoryList(pluginDirectories: List<PluginDirectoryModel>?) {
        if (pluginDirectories == null) {
            return
        }
        WellSql.insert(pluginDirectories).asSingleTransaction(true).execute()
    }

    @JvmStatic
    fun getLastRequestedPageForDirectoryType(directoryType: PluginDirectoryType?): Int {
        val list = getPluginDirectoriesForType(directoryType)
        var page = 0
        list.forEach { page = max(it.page, page) }
        return page
    }

    @JvmStatic
    private fun getPluginDirectoriesForType(
        directoryType: PluginDirectoryType?
    ): List<PluginDirectoryModel> =
        WellSql.select(PluginDirectoryModel::class.java)
            .where()
            .equals(PluginDirectoryModelTable.DIRECTORY_TYPE, directoryType)
            .endWhere()
            .asModel
}
